### PR TITLE
Add OTEL config to Jaeger CR

### LINKED ISSUE #1004: Use Jaeger-OTEL collector in jaeger-operator
Is our aim to swap jaeger-collector with jaeger-opentelemetry-collector in Jaeger CR (e.g. just change the collector image in https://godoc.org/github.com/jaegertracing/jaeger-operator/pkg/apis/jaegertracing/v1#JaegerCollectorSpec)? If so the main problem is configuration otel-collector is configured via configuration file and storage in Jaeger CR is configured by flags - more in https://github.com/jaegertracing/jaeger-opentelemetry-collector/issues/7

The other shade area is the relationship between jaeger-operator and otel-operator (https://github.com/open-telemetry/opentelemetry-operator). For instance if jaeger-operator will start using jaeger-otel-collector (presumably in agent and collector mode) there might not be a need to integrate with otel-collector. Alternatively, Jaeger operator could deprecate its collector and purely integrate with otel-collector.

cc) @jpkrohling

### PR DESCRIPTION
**Pull request by pavolloffay**
Resolves #1004 

Notes:
* If the OTEL config is used with legacy Jaeger images it will cause errors - unknown flag.

A simple CR to test this PR. The collector should successfully deploy and it should use 14269 as HC port. 

```yaml
# setup an elasticsearch with `make es`
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: simple-prod
spec:
  strategy: production
  collector:
    image: jaegertracing/jaeger-opentelemetry-collector:latest
    config:
      processors:
        queued_retry: {}
      extensions:
        health_check:
          port: 14269
      service:
        pipelines:
          traces:
            processors: [queued_retry]
  storage:
    type: elasticsearch
    options:
      es:
        server-urls: http://elasticsearch:9200
```

### CODE REVIEW

File: pkg/config/otelconfig/otelconfig.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s-otel-config", jaeger.Name, component),
+			Namespace: jaeger.Namespace,
+			Labels:    util.Labels(fmt.Sprintf("%s-%s-otel-config", jaeger.Name, component), fmt.Sprintf("%s-otel-config", component), *jaeger),
+			OwnerReferences: []metav1.OwnerReference{
+				util.AsOwner(jaeger),
+			},
+		},
+		Data: map[string]string{
+			configMapKey: string(cfgYml),
```
Comment (rubenvp8510) [inline pkg/config/otelconfig/otelconfig.go:91]: If marshal fails on line 69  this will be an empty configmap, right?

Comment (pavolloffay) [reply pkg/config/otelconfig/otelconfig.go:91]: good point, I think I would rather log an error and do not create anything.

File: pkg/config/otelconfig/otelconfig_test.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		},
+		{
+			expected: true,
+			opts:     v1.NewOptions(map[string]interface{}{"jaeger-config": "/etc/config.yaml"}),
+			otelCfg:  v1.NewFreeForm(map[string]interface{}{"foo": "bar"}),
+		},
+		{
+			expected: false,
+			opts:     v1.NewOptions(map[string]interface{}{}),
+		},
```
Comment (objectiser) [inline pkg/config/otelconfig/otelconfig_test.go:35]: Test case with just `otelCfg`.

File: pkg/config/otelconfig/otelconfig.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	configFlagName     = "--config="
+	configFileLocation = "/etc/jaeger/otel/"
+	configFlagWithFile = configFlagName + configFileLocation + configFileName
+	configMapKey       = "config"
+)
+
+// ShouldCreate returns true if the OTEL config should be created.
+func ShouldCreate(opts v1.Options, otelCfg map[string]interface{}) bool {
+	if _, exists := opts.Map()["config"]; exists {
+		return false
```
Comment (objectiser) [inline pkg/config/otelconfig/otelconfig.go:26]: Should we log a warning if otelcfg also provided, to say will be ignored?

File: pkg/config/otelconfig/otelconfig_test.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	tests := []struct {
+		opts     v1.Options
+		otelCfg  v1.FreeForm
+		expected bool
+	}{
+		{
+			expected: false,
+			opts:     v1.NewOptions(map[string]interface{}{"config": "/etc/config.yaml"}),
+		},
+		{
```
Comment (objectiser) [inline pkg/config/otelconfig/otelconfig_test.go:27]: Another like this, but with `config` - and `expected: false`.

File: pkg/deployment/agent.go [hunk 5]
```diff
@@ -76,6 +78,16 @@ func (a *Agent) Get() *appsv1.DaemonSet {
 
 	commonSpec := util.Merge([]v1.JaegerCommonSpec{a.jaeger.Spec.Agent.JaegerCommonSpec, a.jaeger.Spec.JaegerCommonSpec, baseCommonSpec})
 
+	otelConfg, err := a.jaeger.Spec.Agent.Config.GetMap()
+	if err != nil {
+		a.jaeger.Logger().WithField("error", err).
+			Errorf("Could not parse OTEL config, config map will not be created")
+	} else {
+		if otelconfig.ShouldCreate(a.jaeger.Spec.Agent.Options, otelConfg) {
```
Comment (objectiser) [inline pkg/deployment/agent.go:86]: nit: `else if` (same for other components)

File: pkg/deployment/collector.go [hunk 6]
```diff
@@ -76,6 +78,16 @@ func (c *Collector) Get() *appsv1.Deployment {
 	sampling.Update(c.jaeger, commonSpec, &options)
 	tls.Update(c.jaeger, commonSpec, &options)
 
+	otelConfg, err := c.jaeger.Spec.Collector.Config.GetMap()
+	if err != nil {
+		c.jaeger.Logger().WithField("error", err).
```
Comment (objectiser) [inline pkg/deployment/collector.go:83]: add `component`, as if other components also have otelconfig would be hard to distinguish. (same comment for other components).

File: pkg/config/otelconfig/otelconfig_test.go (hunk 4 above)
Comment (pavolloffay) [reply pkg/config/otelconfig/otelconfig_test.go:27]: I am not sure what you mean. I have added some case, I think there are all the combinations now.

File: pkg/inject/sidecar.go [hunk 7]
```diff
@@ -188,10 +190,20 @@ func container(jaeger *v1.Jaeger, dep *appsv1.Deployment) corev1.Container {
...
+			Errorf("Could not parse OTEL config, config map will not be created")
+	} else if otelconfig.ShouldCreate(jaeger, jaeger.Spec.Agent.Options, otelConfg) {
+		otelconfig.Update(jaeger, "agent", commonSpec, &args)
+	}
+
 	// ensure we have a consistent order of the arguments
 	// see https://github.com/jaegertracing/jaeger-operator/issues/334
 	sort.Strings(args)
 
+	dep.Spec.Template.Spec.Volumes = append(dep.Spec.Template.Spec.Volumes, commonSpec.Volumes...)
```
Comment (pavolloffay) [inline pkg/inject/sidecar.go:206]: I had to add volumes to the pod spec and mounts to agent sidecar container

File: pkg/deployment/agent.go [hunk 8]
```diff
@@ -163,8 +174,10 @@ func (a *Agent) Get() *appsv1.DaemonSet {
 							},
 							InitialDelaySeconds: 1,
 						},
-						Resources: commonSpec.Resources,
+						Resources:    commonSpec.Resources,
+						VolumeMounts: commonSpec.VolumeMounts,
 					}},
+					Volumes:            commonSpec.Volumes,
```
Comment (pavolloffay) [inline pkg/deployment/agent.go:180]: The same as for sidecar, I had to add volumes and mounts to the agent container.

**Comment by pavolloffay** [comment]
All tests passed, the coverage is red, however when I dig in into codecov UI I cannot find the uncovered lines.

**Comment by romil-punetha** [comment]
@pavolloffay 
Any documentation on what all key-values does `config` and `service` section accept?

**Comment by pavolloffay** [comment]
It's OTEL config you can find it in https://github.com/open-telemetry/opentelemetry-collector

**Pull request merged**